File: go-chi/internal/routes/params.go
package routes

import (
	"bufio"
	"cmp"
	"encoding/json/v2"
	"io"
	"net/http"
	"slices"
	"strconv"
	"strings"
	"unicode/utf8"

	"chi-server/internal/utils"

	"github.com/go-chi/chi/v5"
)

const (
	maxFileBytes = 1 << 20 // 1MB
	sniffLen     = 512
	nullByte     = 0x00
)

func RegisterParams(r chi.Router) {
	r.Get("/search", handleSearchParams)
	r.Get("/url/{dynamic}", handleUrlParams)
	r.Get("/header", handleHeaderParams)
	r.Post("/body", handleBodyParams)
	r.Get("/cookie", handleCookieParams)
	r.Post("/form", handleFormParams)
	r.Post("/file", handleFileParams)
}

func handleSearchParams(w http.ResponseWriter, r *http.Request) {
	query := r.URL.Query()

	q := cmp.Or(strings.TrimSpace(query.Get("q")), "none")

	limit := 10
	limitStr := query.Get("limit")
	if limitStr != "" && !strings.Contains(limitStr, ".") {
		if n, err := strconv.ParseInt(limitStr, 10, 64); err == nil && n >= -(1<<53-1) && n <= (1<<53-1) {
			limit = int(n)
		}
	}

	utils.WriteResponse(w, http.StatusOK, map[string]any{"search": q, "limit": limit})
}

func handleUrlParams(w http.ResponseWriter, r *http.Request) {
	dynamic := chi.URLParam(r, "dynamic")
	utils.WriteResponse(w, http.StatusOK, map[string]any{"dynamic": dynamic})
}

func handleHeaderParams(w http.ResponseWriter, r *http.Request) {
	header := cmp.Or(strings.TrimSpace(r.Header.Get("X-Custom-Header")), "none")
	utils.WriteResponse(w, http.StatusOK, map[string]any{"header": header})
}

func handleBodyParams(w http.ResponseWriter, r *http.Request) {
	defer r.Body.Close()

	var body map[string]any
	if err := json.UnmarshalRead(r.Body, &body); err != nil {
		utils.WriteError(w, http.StatusBadRequest, "invalid JSON body", nil)
		return
	}

	utils.WriteResponse(w, http.StatusOK, map[string]any{"body": body})
}

func handleCookieParams(w http.ResponseWriter, r *http.Request) {
	cookieStr, err := r.Cookie("foo")

	cookie := "none"
	if trimmed := strings.TrimSpace(cookieStr.Value); err == nil && trimmed != "" {
		cookie = trimmed
	}

	http.SetCookie(w, &http.Cookie{Name: "bar", Value: "12345", MaxAge: 10, HttpOnly: true, Path: "/"})
	utils.WriteResponse(w, http.StatusOK, map[string]any{"cookie": cookie})
}

func handleFormParams(w http.ResponseWriter, r *http.Request) {
	defer r.Body.Close()

	contentType := strings.ToLower(r.Header.Get("Content-Type"))
	if !strings.HasPrefix(contentType, "application/x-www-form-urlencoded") && !strings.HasPrefix(contentType, "multipart/form-data") {
		utils.WriteError(w, http.StatusBadRequest, "invalid form data", nil)
		return
	}

	if err := r.ParseForm(); err != nil {
		utils.WriteError(w, http.StatusBadRequest, "invalid form data", nil)
		return
	}

	name := cmp.Or(strings.TrimSpace(r.FormValue("name")), "none")

	age := 0
	ageStr := strings.TrimSpace(r.FormValue("age"))
	if ageStr != "" && !strings.Contains(ageStr, ".") {
		if n, err := strconv.ParseInt(ageStr, 10, 64); err == nil && n >= -(1<<53-1) && n <= (1<<53-1) {
			age = int(n)
		}
	}

	utils.WriteResponse(w, http.StatusOK, map[string]any{"name": name, "age": age})
}

func handleFileParams(w http.ResponseWriter, r *http.Request) {
	defer r.Body.Close()

	contentType := strings.ToLower(r.Header.Get("Content-Type"))
	if !strings.HasPrefix(contentType, "multipart/form-data") {
		utils.WriteError(w, http.StatusBadRequest, "invalid multipart form data", nil)
		return
	}

	if err := r.ParseMultipartForm(maxFileBytes); err != nil {
		utils.WriteError(w, http.StatusBadRequest, "invalid multipart form data", nil)
		return
	}

	file, fileHeader, err := r.FormFile("file")
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "file not found in form data", nil)
		return
	}
	defer file.Close()

	br := bufio.NewReader(file)

	head, err := br.Peek(sniffLen)
	if err != nil && err != io.EOF {
		utils.WriteError(w, http.StatusBadRequest, "unable to read file", nil)
		return
	}

	if mime := http.DetectContentType(head); !strings.HasPrefix(mime, "text/plain") {
		utils.WriteError(w, http.StatusUnsupportedMediaType, "only text/plain files are allowed", nil)
		return
	}

	if slices.Contains(head, nullByte) {
		utils.WriteError(w, http.StatusUnsupportedMediaType, "file does not look like plain text", nil)
		return
	}

	limited := io.LimitReader(br, maxFileBytes+1)
	data, err := io.ReadAll(limited)
	if err != nil {
		utils.WriteError(w, http.StatusBadRequest, "unable to read file content", nil)
		return
	}
	if int64(len(data)) > maxFileBytes {
		utils.WriteError(w, http.StatusRequestEntityTooLarge, "file size exceeds limit", nil)
		return
	}
	if slices.Contains(data, nullByte) {
		utils.WriteError(w, http.StatusUnsupportedMediaType, "file does not look like plain text", nil)
		return
	}
	if !utf8.Valid(data) {
		utils.WriteError(w, http.StatusUnsupportedMediaType, "file does not look like plain text", nil)
		return
	}

	utils.WriteResponse(w, http.StatusOK, map[string]any{
		"filename": fileHeader.Filename,
		"size":     len(data),
		"content":  string(data),
	})
}

File: node-express/src/routes/params.ts
import express, { type NextFunction, type Request, type Response, type Router } from "express";
import multer from "multer";

const MAX_REQUEST_BYTES = 10 * 1024 * 1024; // 10 MB
const MAX_FILE_BYTES = 1 << 20; // 1MB
const SNIFF_LEN = 512;
const NULL_BYTE = 0x00;

const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: MAX_FILE_BYTES },
});

const formParser = multer({
  limits: { fieldSize: MAX_REQUEST_BYTES },
}).none();

export const paramsRouter: Router = express.Router();

paramsRouter.get("/search", (req: Request, res: Response) => {
  const qValue = Array.isArray(req.query.q) ? req.query.q[0] : req.query.q;
  const q = typeof qValue === "string" && qValue.trim() !== "" ? qValue.trim() : "none";

  const limitValue = Array.isArray(req.query.limit) ? req.query.limit[0] : req.query.limit;
  const limitStr = typeof limitValue === "string" ? limitValue : undefined;
  const limitNum = Number(limitStr);
  const limit = limitStr !== undefined && !limitStr.includes(".") && Number.isSafeInteger(limitNum) ? limitNum : 10;

  res.json({ search: q, limit });
});

paramsRouter.get("/url/:dynamic", (req: Request, res: Response) => {
  res.json({ dynamic: req.params.dynamic });
});

paramsRouter.get("/header", (req: Request, res: Response) => {
  const header = req.get("X-Custom-Header")?.trim() || "none";
  res.json({ header });
});

paramsRouter.post("/body", (req: Request, res: Response) => {
  const body = req.body;

  if (typeof body !== "object" || body === null || Array.isArray(body)) {
    res.status(400).json({ error: "invalid JSON body" });
    return;
  }

  res.json({ body });
});

paramsRouter.get("/cookie", (req: Request, res: Response) => {
  const cookie = req.cookies?.foo?.trim() || "none";
  res.cookie("bar", "12345", { maxAge: 10_000, httpOnly: true, path: "/" });
  res.json({ cookie });
});

function handleForm(req: Request, res: Response) {
  const body = req.body;
  if (typeof body !== "object" || body === null || Array.isArray(body)) {
    res.status(400).json({ error: "invalid form data" });
    return;
  }

  const name = typeof body.name === "string" ? body.name.trim() || "none" : "none";

  const ageStr = typeof body.age === "string" ? body.age.trim() : "0";
  const ageNum = Number(ageStr);
  const age = !ageStr.includes(".") && Number.isSafeInteger(ageNum) ? ageNum : 0;

  res.json({ name, age });
}

paramsRouter.post("/form", (req: Request, res: Response) => {
  const contentType = req.get("content-type")?.toLowerCase() ?? "";
  if (!contentType.startsWith("application/x-www-form-urlencoded") && !contentType.startsWith("multipart/form-data")) {
    res.status(400).json({ error: "invalid form data" });
    return;
  }

  if (req.is("multipart/form-data")) {
    formParser(req, res, (err?: unknown) => {
      if (err) {
        res.status(400).json({ error: "invalid form data" });
        return;
      }
      handleForm(req, res);
    });
    return;
  }

  handleForm(req, res);
});

paramsRouter.post("/file", (req: Request, res: Response, next: NextFunction) => {
  const contentType = req.get("content-type")?.toLowerCase() ?? "";
  if (!contentType.startsWith("multipart/form-data")) {
    res.status(400).json({ error: "invalid multipart form data" });
    return;
  }

  upload.single("file")(req, res, (err?: unknown) => {
    if (err) {
      next(err);
      return;
    }

    const file = req.file;
    if (!file) {
      res.status(400).json({ error: "file not found in form data" });
      return;
    }

    if (!file.mimetype || !file.mimetype.startsWith("text/plain")) {
      res.status(415).json({ error: "only text/plain files are allowed" });
      return;
    }

    if (file.size > MAX_FILE_BYTES) {
      res.status(413).json({ error: "file size exceeds limit" });
      return;
    }

    const head = file.buffer.subarray(0, SNIFF_LEN);
    if (head.includes(NULL_BYTE)) {
      res.status(415).json({ error: "file does not look like plain text" });
      return;
    }

    if (file.buffer.includes(NULL_BYTE)) {
      res.status(415).json({ error: "file does not look like plain text" });
      return;
    }

    let content: string;
    try {
      const decoder = new TextDecoder("utf-8", { fatal: true });
      content = decoder.decode(file.buffer);
    } catch {
      res.status(415).json({ error: "file does not look like plain text" });
      return;
    }

    res.json({
      filename: file.originalname,
      size: file.buffer.length,
      content,
    });
  });
});
